dags/trending-snapshots-dag.py

```diff
@@ -0,0 +1,166 @@
+from airflow.providers.snowflake.hooks.snowflake import SnowflakeHook
+from orca.services.synapse import SynapseHook
+
+
+SYNAPSE_RESULTS_TABLE = "syn61932294"
+
+dag_params = {
+    "snowflake_conn_id": Param("SNOWFLAKE_SYSADMIN_PORTAL_RAW_CONN", type="string"),
+    "synapse_conn_id": Param("SYNAPSE_ORCA_SERVICE_ACCOUNT_CONN", type="string"),
+    "current_date": Param(date.today().strftime("%Y-%m-%d"), type="string"),
+    "time_window": Param("30d", type="string"),
+    }
+
+dag_config = {
+    # run on the 1st of every month at midnight
+    "schedule_interval": "0 0 1 * *",
+    "start_date": datetime(2024, 7, 1),
+    "catchup": False,
+    "default_args": {
+        "retries": 1,
+    },
+    "tags": ["snowflake"],
+    "params": dag_params,
+}
+
+
+@dataclass
+class SnapshotsMetric:
+    """
+    Dataclass to hold the download metrics from Synapse.
+
+    Attributes:
+        total_data_size_in_pib: The size of the data hosted on Synapse in PiB
+        active_users_last_month: The number of active users last month
+        total_downloads_last_month: The total number of downloads by users last month
+    """
+
+    project_id: float
+    n_unique_users: int
+    last_download_date: int
+    total_data_size_in_gib: float
+
+
+@dag(**dag_config)
+def trending_snapshots() -> None:
+    """
+    """
+
+    @task
+    def get_trending_snapshots(**context) -> List[SnapshotsMetric]:
+        """Execute the query on Snowflake and return the results."""
+        snow_hook = SnowflakeHook(context["params"]["snowflake_conn_id"])
+        ctx = snow_hook.get_conn()
+        cs = ctx.cursor()
+        query = f"""
+                WITH RECENT_DOWNLOADS AS (
+                    SELECT *
+                    FROM SYNAPSE_DATA_WAREHOUSE.SYNAPSE.FILEDOWNLOAD
+                    WHERE 1=1
+                    AND RECORD_DATE > DATEADD(DAY, -28, '2024-06-26') 
+                    AND RECORD_DATE <= '2024-06-26'
+                    AND STACK = 'prod'
+                    AND PROJECT_ID IS NOT NULL
+                ),
+                PUBLIC_PROJECTS AS (
+                    SELECT NAME, PROJECT_ID
+                    FROM SYNAPSE_DATA_WAREHOUSE.SYNAPSE.NODE_LATEST
+                    WHERE 1=1
+                    AND NODE_TYPE = 'project'
+                    AND IS_PUBLIC = TRUE
+                )
+
+                SELECT 
+                    recent_downloads.PROJECT_ID, 
+                    COUNT(DISTINCT recent_downloads.USER_ID) AS unique_users,
+                    MAX(recent_downloads.RECORD_DATE) AS last_download_date,
+                    SUM(file_latest.CONTENT_SIZE) / POWER(1024, 3) AS TOTAL_DATA_SIZE_IN_GIB
+
+                -- Select the table you want to curate from
+                FROM 
+                    RECENT_DOWNLOADS recent_downloads
+
+                -- Join FILE_LATEST
+                JOIN 
+                    SYNAPSE_DATA_WAREHOUSE.SYNAPSE.FILE_LATEST file_latest
+                ON 
```

Comment:
What is the purpose of this join?

Reply:
That makes sense, but I think you are joining against the filedownload table which would give you total data transferred in bytes